jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppClassLoader.java

```diff
@@ -319,29 +319,37 @@ public void addJars(Resource lib)
     {
         if (lib.exists() && lib.isDirectory())
         {
-            String[] files = lib.list();
-            if (files != null)
+            String[] entries = lib.list();
+            if (entries != null)
             {
-                Arrays.sort(files);
-            }
-            for (int f = 0; files != null && f < files.length; f++)
-            {
-                try
+                Arrays.sort(entries);
+
+                for (String entry : entries)
                 {
-                    Resource fn = lib.addPath(files[f]);
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("addJar - {}", fn);
-                    String fnlc = fn.getName().toLowerCase(Locale.ENGLISH);
-                    // don't check if this is a directory (prevents use of symlinks), see Bug 353165
-                    if (isFileSupported(fnlc))
+                    try
                     {
-                        String jar = URIUtil.encodeSpecific(fn.toString(), ",;");
-                        addClassPath(jar);
+                        Resource resource = lib.addPath(entry);
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("addJar - {}", resource);
+                        if (resource.isDirectory())
+                        {
+                            addURL(resource.getURI().toURL());
```

Comment:
Wwe've ever supported an extraClasspath directory hierarchy that is a mix of classes dirs and jar files. The arg to addJars() has always been interpreted as a dir that contains jar/zip files, and only the jar/zip files were added to the classpath. It looks like line 334 is going to add the subdirs of the lib to the classpath, but that's never been the accepted behaviour.

Reply:
No, it doesn't add sub-directories, it only adds the directory itself as a directory reference.
Similar to how WebInfConfiguration.findExtraClassPathDirs() works with glob.